File: ocuk2010/c/ocuk-stage1-05.c
// Steve Phillips
// 2010.12.13
// OCUK Stage 1

#include <stdio.h>

int main(void) {

    int MAX_NUM = 1500000;

    unsigned int chain_length, num, number;
    unsigned int longest_chain = 0;
    unsigned int longest_n = 0;

    for (num = 2; num < MAX_NUM; num++) {
	number = num;
	chain_length = 0;

	while (number != 1) {
  	    if (!(number & 1)) { // Even
		number >>= 1;
	    } else {             // Odd
		number += (number << 1) + 1;
	    }
	    chain_length += 1;

	    if (chain_length > longest_chain) {
		longest_chain = chain_length;
		longest_n = num;
	    }
	}
    }

    printf("%d", longest_n);
    //printf("n = %d generates chain length %d", longest_n, longest_chain);
    return 0;
}

File: ocuk2010/go/ocuk-stage1-05.go
// Steve Phillips
// 2010.12.13
// OCUK Stage 1

package main

import (
	"fmt"
)

func main() {

	//MAX_NUM := 100
	MAX_NUM := 1500000

	longest_chain, longest_n := 0, 0

	for num := 2; num < MAX_NUM; num++ {
		number := num
		chain_length := 0

		for number != 1 {
			if (number & 1) == 0 { // Even
				number >>= 1
			} else {               // Odd
				number += (number << 1) + 1
			}
			chain_length += 1

			if chain_length > longest_chain {
				longest_chain = chain_length
				longest_n = num
			}
		}
	}

	fmt.Printf("n = %d generates chain length %d", longest_n, longest_chain)
}